### ISSUE TITLE
Base class with parametrized constructor

### ISSUE THREAD

**Issue by MichalPavlik** [RazorGenerator/RazorGenerator#17]
Hi, RazorGenerator is great but in our project we would like to use "@ inherits" directive with class without default constructor (Generator: Template). It's possible to generate appropriate constructors directly into generated C# classes by RazorGenerator custom tool? We have a lot of templates and create partial classes only for parametrized constructor is really mayhem. Thanks.

**Comment by davidebbo** [comment]
Had not tried that, but I came up with a [sample](https://github.com/davidebbo-test/RazorGeneratorSimpleTemplate/blob/master/RazorGeneratorSimpleTemplate/SampleTemplateWithBase.cshtml) that seems to work.

At design time, VS is a little confused because the designer doesn't know the name of the generated class, so I think it's not recognizing it as a ctor. But other than that, it generates the right thing and works.

Is that what you were after?

**Comment by MichalPavlik** [comment]
Yes it works. I tried this before but I didn't even try to compile it when I saw red squiggles :)

But I believe you can generate constructor(s) to "<name>.generated.cs" class automatically via reflection in your T4 template. It will be even better to have this functionality out of the box without boilerplate imperative code in cshtml views :)

**Comment by MichalPavlik** [comment]
So I downloaded source and implemented some code into TemplateTransformer.ProcessGeneratedCode.

....
if (generatedClass.BaseTypes.Count > 0)
            {
                string typeName = generatedClass.BaseTypes[0].BaseType;

                Type baseType = Type.GetType(typeName) ?? 
                    AppDomain.CurrentDomain.GetAssemblies()
                        .Select(a => a.GetType(typeName))
                        .FirstOrDefault(t => t != null);

                if (baseType != null)
                {
                    foreach (ConstructorInfo constructorInfo in baseType.GetConstructors().Where(c => c.IsPublic))
                    {
                        CodeConstructor constructor = new CodeConstructor()
                        {
                            Attributes = MemberAttributes.Public
                        };

                        foreach(ParameterInfo parameterInfo in constructorInfo.GetParameters())
                        {
                            constructor.Parameters.Add(new CodeParameterDeclarationExpression(parameterInfo.ParameterType.FullName, parameterInfo.Name));
                            constructor.BaseConstructorArgs.Add(new CodeArgumentReferenceExpression(parameterInfo.Name));
                        }

                        generatedClass.Members.Add(constructor);
                    }
                }
            }
....

It seems to be working (I checked test output only by my eyes :)) The problem might be in runtime type discovery, because assembly with parent type must be loaded in executing app domain. You can use this implementation as bootstrap if you want implement this functionality:)

**Comment by davidebbo** [comment]
Feel free to send a PR if you get something working. That's how it's done :)

So your goal is to always generate a passthrough ctor?

For runtime type discovery, it may be better to use the VS code model instead of reflection. e.g. that's what we do in [T4MVC](https://github.com/T4MVC/T4MVC).

**Comment by MichalPavlik** [comment]
I never worked with Git (my employer is corporation with on-prem TFS) but I will try :)

Yes, we need empty passthrough ctors without logic.

The VS API might be better :)

And one more thing... It's possible to implement custom IRazorCodeTransformer in our project codebase and use it for class generator without copying assembly with transformer outside our solution?

**Comment by JoshMcCullough** [comment]
How do you define which template RazorGenerator should use? Can I create a template for my project and have all generated views use it instead of the default template?
